### DISCUSSION TITLE
association_proxy filter out None values

### DISCUSSION THREAD

**Discussion by alex-frx** [sqlalchemy/sqlalchemy#8229] [Usage Questions]
Hello, 
im Using an `association_proxy` like this: 
``` python
study_participantions = association_proxy("quests", "study_participant",creator=lambda sp: sp.questionnaire)
```
I my Db there is:
 - **A Patient table**
 - **A StudyParticipant table**
 - **A Questionnaire table**
 - ** A Study Table **

**Patient** and **Questionnaire** are a **Many-To-One relationship**
A **Questionnaire** can be Part of a **StudyParticiapnt** via a **One-To-One** relationship
StudyParticipant and Patient are not directly linked since StudyParticipant can be annonymous.
So via getter and setter i can query the Patient trough the questionnaire.

Since im working with an existing codebase I have to keep the patient inside the questionnaire

The StudyParticipant can be find via the proxy. getting and setting works but if the Questionnaire is not part of a StudyParticiapnt the returned array contains `None` values is it possible to filter them out so i would get a clean array? For sure it should still be an 
`sqlalchemy.ext.associationproxy._AssociationList` so appending and removing to it would still work. 

Simplified Classes:

``` python

from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker
from sqlalchemy.ext.associationproxy import association_proxy
from sqlalchemy import (
    Column, Table, ForeignKey, Sequence, Integer, String, Text, Boolean,
    DateTime, Date, Enum, column, event, UniqueConstraint
)
from sqlalchemy.schema import DDL
from sqlalchemy.orm import relationship, backref, make_transient, reconstructor
from sqlalchemy.ext.declarative import declarative_base, DeclarativeMeta
from sqlalchemy.ext.associationproxy import association_proxy
from sqlalchemy.ext.hybrid import hybrid_property


model_registry = dict()
Model = declarative_base(class_registry=model_registry)


class Patient(Model):
    __tablename__ = 'patient'

    id = Column(Integer, primary_key=True)
    
    study_participations = association_proxy(
        "quests", "study_participant", creator=lambda sp: sp.questionnaire)

    


class StudyParticipant(Model):
    __tablename__ = "study_participant"

    id = Column(Integer, primary_key=True)
    pseudonym = Column(String(40), nullable=True)

    questionnaire = relationship(
        "Questionnaire", backref="study_participant", uselist=False)
    study_id = Column(Integer, ForeignKey('study.id'), nullable=False)


    @hybrid_property
    def patient(self):
        # plus sanity check if study in anonymous for clearity deleted the study relationshio
        return self.questionnaire.patient

    @patient.setter
    def patient(self, patient):
        # plus sanity check if study in anonymous for clearity deleted the study relationshio
        self.questionnaire.patient = patient


class Questionnaire(Model):
    __tablename__ = 'questionnaire'

    id = Column(Integer, primary_key=True)
    
    @hybrid_property
    def is_part_of_study(self):
        return self.study_participant is not None

    @hybrid_property
    def study(self):
        if self.is_part_of_study:
            return self.study_participant.study
        else:
            return None

    @study.setter
    def study(self, study):
        if self.is_part_of_study:
            self.study_participant.study = study
        else:
            raise ValueError("Cannot set study for questionnaire not part of a study")
    patient_id = Column(Integer, ForeignKey('patient.id'), nullable=False)
    patient = relationship('Patient', backref='quests')
    study_participant_id = Column(Integer, ForeignKey('study_participant.id'), nullable=True) 
   


class Study(Model):
    __tablename__ = "study"

    id = Column(Integer, primary_key=True)
    participants = relationship("StudyParticipant", backref="study")


engine = create_engine('sqlite:///:memory:')
Session = sessionmaker(bind=engine)
session = Session()

Model.metadata.create_all(engine)

p1 = Patient()
s = Study()
q = Questionnaire(patient=p1)
session.commit()
sp = StudyParticipant(questionnaire=q, study=s)
session.commit()
# assigne study and questionnaire to patient
session.commit()
print(p1.study_participations)


p2 = Patient()
p2.study_participations.append(sp)
session.commit()
print(p2.study_participations) # now he is part of the study
print(p1.study_participations) # he got removed from the study

sp.questionnaire.patient = p1
session.commit()
print(p1.study_participations) # now he is part of the study again

```

**Comment by zzzeek** [comment]
i dont see a way to get what you want directly without changing how the relationships work, or going through the task of creating a subclass of `_AssociationList`, which can be done.     I'm not seeing how StudyParticipant joins in and this looks a bit like maybe I'd connect things differently, or not use assocaition proxy, not sure.

**Comment by alex-frx** [comment]
@zzzeek 
I fixed the example and added a few lines so it is easier to get a gasp how the realtionship works.
Maybe it is an x-y question and im looking at the wrong spot.
But why i want this to work in the first place is to be able to easy work with the objects like this

``` python 
query = session.query
p1 = query(models.Patient).first(1)
s = query(models.Study).first()
q = query(models.Questionnaire).first()

sp = models.StudyParticipant(questionnaire=q, study=s)
p1.study_particpations.append(sp) #assigne study and questionnaire to patient

p2 = query(models.Patient).first.get(2)
p2.study_participantions.append(sp) # this effectly swapws the questionnaire to the patient
sp.questionnaire.patient = p1 # this effectly swaps the patient to the questionnaire back so it is interchangeable

```

If I put a realtionship a direct relation between `Patient` and `StudyParticipant` i have to update at tow points. So i can just update at some point and everything is fine

**Reply by zzzeek** [reply]
`questionaiire_patient_join`, not in the example.  I have no way to see how this works relationally

if you can please format a fully working [MCVE](http://stackoverflow.com/help/mcve) thanks

**Reply by alex-frx** [reply]
@zzzeek Sorry for the late response. I updated the Question to contain a fully working MCVE. 

I simplified the RelationShip between Questionnaire and Patien. It would just do some sanity checks and throw Errors Im pretty sure it is not important for my question basiccly it just does this `join = (Patient.id == Questionnaire.patient_id)`

**Reply by zzzeek** [reply]
your test case prints:

```
[<__main__.StudyParticipant object at 0x7f8447d77340>]
[<__main__.StudyParticipant object at 0x7f8447d77340>]
[]
[<__main__.StudyParticipant object at 0x7f8447d77340>]

```

which is because as you move the studyparticipant between p1 and p2, you are also moving the Questionnaire object.   so, that works?  if that's how you are using the objects here then you don't get None in the list.     Whats the case for Questionnaire.study_participant_id = NULL ?
